bitcoin/src/taproot.rs

```diff
@@ -577,23 +792,129 @@ impl NodeInfo {
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::SeqAccess<'de>,
+            {
+                let mut builder = TaprootBuilder::new();
+                while let Some(depth) = seq.next_element()? {
+                    let tap_leaf : TapLeaf = seq
+                        .next_element()?
+                        .ok_or_else(|| serde::de::Error::custom("Missing tap_leaf"))?;
+                    match tap_leaf {
+                        TapLeaf::Script(script, ver) => {
+                            builder = builder.add_leaf_with_ver(depth, script, ver).map_err(|e| {
+                                serde::de::Error::custom(format!("Leaf insertion error: {}", e))
+                            })?;
+                        },
+                        TapLeaf::Hidden(h) => {
+                            builder = builder.add_hidden_node(depth, h).map_err(|e| {
+                                serde::de::Error::custom(format!("Hidden node insertion error: {}", e))
+                            })?;
+                        },
+                    }
+                }
+                NodeInfo::try_from(builder).map_err(|e| {
+                    serde::de::Error::custom(format!("Incomplete taproot tree: {}", e))
+                })
+            }
+        }
+
+        deserializer.deserialize_seq(SeqVisitor)
+    }
+}
+
+/// Leaf node in a taproot tree. Can be either hidden or known.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub enum TapLeaf {
+    /// A known script
+    Script(ScriptBuf, LeafVersion),
+    /// Hidden Node with the given leaf hash
+    Hidden(sha256::Hash), // TODO: This will be TapNodeHash after 1479 is merged.
+}
+
+impl TapLeaf {
+    /// Obtains the hidden leaf hash if the leaf is hidden.
+    pub fn as_hidden(&self) -> Option<&sha256::Hash> {
+        if let Self::Hidden(v) = self {
+            Some(v)
+        } else {
+            None
+        }
+    }
+
+    /// Obtains the script and version if the leaf is known.
+    pub fn as_script(&self) -> Option<(&Script, LeafVersion)> {
+        if let Self::Script(script, ver) = self {
+            Some((script, *ver))
+        } else {
+            None
+        }
+    }
+}
+
 /// Store information about taproot leaf node.
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
 pub struct ScriptLeaf {
-    /// The underlying script.
-    script: ScriptBuf,
-    /// The leaf version.
-    ver: LeafVersion,
+    /// The [`TapLeaf`]
+    pub leaf: TapLeaf,
     /// The merkle proof (hashing partners) to get this node.
-    merkle_branch: TaprootMerkleBranch,
+    pub merkle_branch: TaprootMerkleBranch,
```

Comment:
Looks like outside code could break invariants when this is `pub`?

Reply:
Indeed, it might break invariants. I remember thinking about this while adding `pub`, but I was mistaken :(